flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkCalciteSqlValidator.java

```diff
@@ -321,4 +332,131 @@ protected void addToSelectList(
+                                                            .filter(SqlIdentifier.class::isInstance)
+                                                            .map(SqlIdentifier.class::cast))
+                            .collect(Collectors.toList());
+
+            for (int i = 0; i < call.operandCount(); i++) {
+                final SqlIdentifier tableArg = explicitTableArgs.get(i);
+                if (tableArg != null) {
+                    call.setOperand(i, new ExplicitTableSqlSelect(tableArg, descriptors));
+                }
+            }
+        }
+
+        return rewritten;
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // Column expansion
+    // --------------------------------------------------------------------------------------------
+
+    /**
+     * A special {@link SqlSelect} to capture the origin of a {@link SqlKind#EXPLICIT_TABLE} within
+     * TVF operands.
+     */
+    private static class ExplicitTableSqlSelect extends SqlSelect {
+
+        private final List<SqlIdentifier> descriptors;
+
+        public ExplicitTableSqlSelect(SqlIdentifier table, List<SqlIdentifier> descriptors) {
+            super(
+                    SqlParserPos.ZERO,
+                    null,
+                    SqlNodeList.of(SqlIdentifier.star(SqlParserPos.ZERO)),
+                    table,
+                    null,
+                    null,
+                    null,
+                    null,
+                    null,
+                    null,
+                    null,
+                    null);
+            this.descriptors = descriptors;
+        }
+    }
+
+    /**
+     * Returns whether the given column has been declared in a {@link SqlKind#DESCRIPTOR} next to a
+     * {@link SqlKind#EXPLICIT_TABLE} within TVF operands.
+     */
+    private static boolean declaredDescriptorColumn(SelectScope scope, Column column) {
+        if (!(scope.getNode() instanceof ExplicitTableSqlSelect)) {
+            return false;
+        }
+        final ExplicitTableSqlSelect select = (ExplicitTableSqlSelect) scope.getNode();
+        return select.descriptors.stream()
+                .map(SqlIdentifier::getSimple)
+                .anyMatch(id -> id.equals(column.getName()));
+    }
+
+    /**
+     * Returns all {@link SqlKind#EXPLICIT_TABLE} operands within TVF operands. A list entry is
+     * {@code null} if the operand is not an {@link SqlKind#EXPLICIT_TABLE}.
+     */
+    private static List<SqlIdentifier> getExplicitTableOperands(SqlNode node) {
+        if (!(node instanceof SqlBasicCall)) {
+            return null;
+        }
+        final SqlBasicCall call = (SqlBasicCall) node;
+
+        if (!(call.getOperator() instanceof SqlFunction)) {
+            return null;
+        }
+        final SqlFunction function = (SqlFunction) call.getOperator();
+
+        if (function.getFunctionType() != SqlFunctionCategory.USER_DEFINED_TABLE_FUNCTION) {
```

Comment:
Why are you looking only at UDFs (USER_DEFINED_TABLE_FUNCTION) here?

Reply:
For performance, I want to exit this method early. And currently all PTFs have this catagory.